controller/controller_server/utils.py

```diff
@@ -8,43 +10,161 @@
 import controller.controller_server.messages as messages
 from controller.array_action.config import FC_CONNECTIVITY_TYPE, ISCSI_CONNECTIVITY_TYPE
 from controller.common.csi_logger import get_stdout_logger
+from controller.controller_server.controller_types import ArrayConnectionInfo, ObjectIdInfo, ObjectParameters
 from controller.controller_server.errors import ObjectIdError, ValidationException
 from controller.csi_general import csi_pb2
 
 logger = get_stdout_logger()
 
 
-def get_array_connection_info_from_secret(secrets):
+def _parse_raw_json(raw_json):
+    try:
+        parsed_json = json.loads(raw_json)
+    except json.decoder.JSONDecodeError:
+        raise ValidationException(messages.invalid_json_parameter_message.format(raw_json))
+    return parsed_json
+
+
+def _is_topology_match(system_topologies, node_topologies):
+    for topologies in system_topologies:
+        logger.debug(
+            "Comparing topologies: system topologies: {},"
+            " node topologies: {}".format(topologies, node_topologies))
+        if topologies.items() <= node_topologies.items():
+            return True
+    return False
+
+
+def get_volume_topologies(request):
+    if request.HasField(config.REQUEST_ACCESSIBILITY_REQUIREMENTS_FIELD):
+        accessibility_requirements = request.accessibility_requirements
+        if accessibility_requirements.preferred:
+            topologies = accessibility_requirements.preferred[0].segments
+            logger.info("Chosen volume topologies: {}".format(topologies))
+            return topologies
+
+
+def get_system_info_by_topologies(secrets_config, node_topologies):
+    for system_id, system_info in secrets_config.items():
+        system_topologies = system_info.get(config.SECRET_SUPPORTED_TOPOLOGIES_PARAMETER)
+        if _is_topology_match(system_topologies, node_topologies):
+            return system_info, system_id
+    raise ValidationException(messages.no_way_to_identify_system_from_secrets_config_message)
```

Comment:
this could be more specific, e.g. `no system match requested topology`

Reply:
done